src/Barbies/Internal/DistributiveB.hs

```diff
@@ -0,0 +1,116 @@
+
+import Data.Functor.Compose   (Compose (..))
+import Data.Functor.Identity  (Identity (..))
+import Data.Functor.Product   (Product (..))
+import Data.Generics.GenericN
+import Data.Proxy             (Proxy (..))
+import Data.Distributive
+import Data.Kind              (Type)
+
+-- | A 'FunctorB' where the effects can be distributed to the fields:
+-- `bdistribute` turns an effectful way of making a Barbie-type
+-- into a pure Barbie-type with effectful ways of computing the
+-- values of its fields.
+--
+-- From 'bdistribute', we can define a pure skeleton where very field
+-- is a getter of that field:
+--
+-- @
+-- 'bshape' :: ('DistributiveB' b) => b ((->) (b 'Identity'))
+-- 'bshape' = 'bdistribute'' 'id'
+-- @
+--
+--  It should satisfy the following law
+--
+-- @
+-- 'bmap' ('Identity' . ('$' b)) 'bshape' = b
+-- @
+--
+--  It is to 'FunctorB' in the same way as 'Distributive'
+--  relates to 'Functor'.
+--
+-- There is a default implementation of 'bdistribute' based on
+-- 'Generic'.  Intuitively, it works on product types where the shape
+-- of a pure value is uniquely defined and every field is covered by
+-- the argument @f@.
+class (FunctorB b) => DistributiveB (b :: (k -> Type) -> Type) where
+  bdistribute :: Distributive f => f (b g) -> b (Compose f g)
+  -- bdistribute x = bmap (\f -> Compose $ fmap f . bsequence' <$> x) bshape
+
+  default bdistribute
+    :: forall f g
+    .  CanDeriveDistributiveB b f g
+    => Distributive f => f (b g) -> b (Compose f g)
+  bdistribute = gbdistributeDefault
+
+bshape :: (DistributiveB b) => b ((->) (b Identity))
+bshape = bdistribute' id
+
+-- | A version of `bdistribute` with @g@ specialized to `Identity`.
+bdistribute' :: (DistributiveB b, Distributive f) => f (b Identity) -> b f
+bdistribute' = bmap (fmap runIdentity . getCompose) . bdistribute
+
+type CanDeriveDistributiveB b f g
+  = ( GenericP 0 (b g)
+    , GenericP 0 (b (Compose f g))
+    , GDistributive 0 f (RepP 0 (b g)) (RepP 0 (b (Compose f g)))
+    )
+
+-- | Default implementation of 'bdistribute' based on 'Generic'.
+gbdistributeDefault
+  :: CanDeriveDistributiveB b f g
+  => Distributive f => f (b g) -> b (Compose f g)
+gbdistributeDefault
+  = toP (Proxy @0) . gdistribute (Proxy @0) . fmap (fromP (Proxy @0))
+{-# INLINE gbdistributeDefault #-}
+
+-- ------------------------------------------------------------
+-- Generic derivation: Special cases for FunctorB
+-- -----------------------------------------------------------
+
+type P = Param
+
+instance
+  ( Distributive f
+  , DistributiveB b
+  ) => GDistributive 0 f (Rec (b' (P 0 g)) (b g)) (Rec (b' (P 0 (Compose f g))) (b (Compose f g)))
+  where
+  gdistribute _ = Rec . K1 . bdistribute . fmap (unK1 . unRec)
+  {-# INLINE gdistribute #-}
+
+-- --------------------------------
+-- Instances for base types
+-- --------------------------------
+
+instance DistributiveB Proxy where
+  bdistribute _ = Proxy
+  {-# INLINE bdistribute #-}
```

Comment:
we should add an instance for `Barbies.Unit`

Reply:
Hand-written, or just the generic one?